File: backend/app/services/audio_language.py
"""Detect and classify English / Farsi audio note transcripts."""

from __future__ import annotations

import re
from typing import Literal

AudioNoteLanguage = Literal["en", "fa", "mixed", "unknown"]

_PERSIAN_ARABIC_RE = re.compile(
    r"[\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB50-\uFDFF\uFE70-\uFEFF]"
)
_LATIN_RE = re.compile(r"[A-Za-z]")


def map_whisper_language(code: str | None) -> AudioNoteLanguage:
    if not code:
        return "unknown"
    normalized = code.strip().lower()
    if normalized in {"en", "english"}:
        return "en"
    if normalized in {"fa", "fas", "persian", "farsi"}:
        return "fa"
    return "unknown"


def detect_transcript_language(text: str) -> AudioNoteLanguage:
    stripped = text.strip()
    if not stripped:
        return "unknown"

    has_persian = bool(_PERSIAN_ARABIC_RE.search(stripped))
    has_latin = bool(_LATIN_RE.search(stripped))
    if has_persian and has_latin:
        return "mixed"
    if has_persian:
        return "fa"
    if has_latin:
        return "en"
    return "unknown"


def refine_detected_language(
    text: str,
    whisper_language: str | None = None,
) -> AudioNoteLanguage:
    heuristic = detect_transcript_language(text)
    whisper = map_whisper_language(whisper_language)

    if heuristic == "mixed":
        return "mixed"
    if heuristic in {"en", "fa"} and whisper in {"en", "fa"} and heuristic != whisper:
        return "mixed"
    if heuristic != "unknown":
        return heuristic
    if whisper != "unknown":
        return whisper
    return "unknown"


def language_label(code: AudioNoteLanguage | None) -> str:
    labels = {
        "en": "English",
        "fa": "Farsi / Persian",
        "mixed": "English + Farsi",
        "unknown": "Unknown",
    }
    return labels.get(code or "unknown", "Unknown")

File: frontend/src/lib/audio-language.ts
export type AudioNoteLanguage = "en" | "fa" | "mixed" | "unknown";

const PERSIAN_ARABIC_RE =
  /[\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB50-\uFDFF\uFE70-\uFEFF]/;
const LATIN_RE = /[A-Za-z]/;

export function mapWhisperLanguage(code: string | null | undefined): AudioNoteLanguage {
  if (!code) return "unknown";
  const normalized = code.trim().toLowerCase();
  if (normalized === "en" || normalized === "english") return "en";
  if (normalized === "fa" || normalized === "fas" || normalized === "persian" || normalized === "farsi") {
    return "fa";
  }
  return "unknown";
}

export function detectTranscriptLanguage(text: string): AudioNoteLanguage {
  const stripped = text.trim();
  if (!stripped) return "unknown";

  const hasPersian = PERSIAN_ARABIC_RE.test(stripped);
  const hasLatin = LATIN_RE.test(stripped);
  if (hasPersian && hasLatin) return "mixed";
  if (hasPersian) return "fa";
  if (hasLatin) return "en";
  return "unknown";
}

export function refineDetectedLanguage(
  text: string,
  whisperLanguage?: string | null
): AudioNoteLanguage {
  const heuristic = detectTranscriptLanguage(text);
  const whisper = mapWhisperLanguage(whisperLanguage);

  if (heuristic === "mixed") return "mixed";
  if (heuristic !== "unknown" && whisper !== "unknown" && heuristic !== whisper) {
    return "mixed";
  }
  if (heuristic !== "unknown") return heuristic;
  if (whisper !== "unknown") return whisper;
  return "unknown";
}

export function languageLabel(code: AudioNoteLanguage | null | undefined): string {
  switch (code) {
    case "en":
      return "English";
    case "fa":
      return "Farsi / Persian";
    case "mixed":
      return "English + Farsi";
    default:
      return "Unknown";
  }
}

// Test-friendly camelCase aliases matching backend naming in specs.
export const detect_transcript_language = detectTranscriptLanguage;
export const map_whisper_language = mapWhisperLanguage;
export const refine_detected_language = refineDetectedLanguage;
